lib/global_uid/allocator.rb

```diff
@@ -0,0 +1,66 @@
+module GlobalUid
+  class Allocator
+    attr_reader :allocations, :max_size, :incrementing_by, :connection
+
+    def initialize(max_size: 5, incrementing_by:, connection:)
+      @allocations = []
+      @max_size = max_size
+      @incrementing_by = incrementing_by
+      @connection = connection
+      validate_connection_increment(connection)
+    end
+
+    def add(table)
+      identifier = connection.insert("REPLACE INTO #{table} (stub) VALUES ('a')")
+      increment!(identifier)
+    end
+
+    def add_many(table, count:)
+      increment_by = validate_connection_increment(connection)
+
+      start_id = connection.insert("REPLACE INTO #{table} (stub) VALUES " + (["('a')"] * count).join(','))
+      identifiers = start_id.step(start_id + (count-1) * increment_by, increment_by).to_a
+      identifiers.each { |identifier| increment!(identifier) }
+      identifiers
+    end
+
+    private
+
+    def increment!(identifier)
+      allocations.shift if allocations.size >= max_size
+      allocations << identifier
+
+      begin
```

Comment:
This `begin` block isn't super readable and was revisited in #70 - if we don't want the changes from #70 I can look at teasing out some of it and merging it into this PR

Reply:
Couldn’t we get rid of the begin/end block by checking `suppress_increment_exceptions` before raising the exception:
```rb
unless valid_increment?
  if GlobalUid::Base.global_uid_options[:suppress_increment_exceptions]
    GlobalUid::Base.notify(e, e.message)
    identifier
  else
    raise InvalidIncrementException, \
      "Recently allocated IDs: #{allocations} on '#{connection.current_database}'"
  end
end
```
Of course, `e` and `e.message` would need to be populated somehow…

Reply:
I improved this in #70 - I could pull that in or we could wait for that to be merged
https://github.com/zendesk/global_uid/blob/1f55dd58873784238d792c5a90933a7fdb9f2bc9/lib/global_uid/allocator.rb#L13-L59

Reply:
Let’s wait.

Reply:
I cleaned it up, I was being lazy and should have done it from the start.